test/plugin/test_in_tail.rb

```diff
@@ -345,53 +345,109 @@ def test_emit_with_read_lines_limit(data)
+          end
+          d = create_driver(config)
+          msg = 'test' * 2000 # in_tail reads 8192 bytes at once.
 
-        # We should not do shutdown here due to hard timeout.
-        d.run(expect_emits: 2, shutdown: false) do
           File.open("#{TMP_DIR}/tail.txt", "ab") {|f|
-            for _x in 0..30
+            100.times do
               f.puts msg
             end
           }
+          start_time = Fluent::Clock.now
+
+          # We should not do shutdown here due to hard timeout.
+          d.run(expect_emits: 2, shutdown: false)
+
+          assert_true(Fluent::Clock.now - start_time > 1)
+          assert_equal(num_events.times.map { {"message" => msg} },
+                       d.events.collect { |event| event[2] })
+
+          # Teardown in_tail plugin instance here.
+          d.instance.shutdown
         end
+      end
 
-        events = d.events
-        assert_true(events.length <= num_events)
-        assert_equal({"message" => msg}, events[0][2])
-        assert_equal({"message" => msg}, events[1][2])
+      sub_test_case "reads_bytes_per_second w/ throttled already" do
+        data("flat 8192 bytes, 2 events"        => [:flat, 100, 8192, 2],
+             "flat #{8192*10} bytes, 20 events"  => [:flat, 100, (8192 * 10), 20],
+             "parse #{8192*4} bytes, 8 events"  => [:parse, 100, (8192 * 4), 8],
+             "parse #{8192*10} bytes, 20 events" => [:parse, 100, (8192 * 10), 20],
+             "flat 8k bytes with unit, 2 events"        => [:flat, 100, "8k", 2],
+             "flat #{8*10}k bytes with unit, 20 events"  => [:flat, 100, "#{8*10}k", 20],
+             "parse #{8*4}k bytes with unit, 8 events"  => [:parse, 100, "#{8*4}k", 8],
+             "parse #{8*10}k bytes with unit, 20 events" => [:parse, 100, "#{8*10}k", 20])
+        def test_emit_with_read_bytes_limit_per_second(data)
+          config_style, limit, limit_bytes, num_events = data
+          case config_style
+          when :flat
+            config = CONFIG_READ_FROM_HEAD + SINGLE_LINE_CONFIG + config_element("", "", { "read_lines_limit" => limit, "read_bytes_limit_per_second" => limit_bytes })
+          when :parse
+            config = CONFIG_READ_FROM_HEAD + config_element("", "", { "read_lines_limit" => limit, "read_bytes_limit_per_second" => limit_bytes }) + PARSE_SINGLE_LINE_CONFIG
+          end
+          d = create_driver(config)
+          msg = 'test' * 2000 # in_tail reads 8192 bytes at once.
+
+          mock.proxy(d.instance).io_handler(anything, anything) do |io_handler|
+            require 'fluent/config/types'
+            limit_bytes_value = Fluent::Config.size_value(limit_bytes)
+            io_handler.instance_variable_set(:@number_bytes_read, limit_bytes_value)
+            mock.proxy(io_handler).handle_notify.at_least(5)
```

Comment:
In Windows and macOS, `handle_inotify` is called 2 times instead of 5.
But I'm not sure why they don't call it 5 times.... :confounded:

Reply:
handle_inotify will be called on every file change at https://github.com/fluent/fluentd/blob/e2e51d1182188d8d498e096dfab1a031dafacb75/test/plugin/test_in_tail.rb#L442-L447

Reply:
Why I choose `at_least(5)` instead of `at_least(9)` is that considering instability of CI.